src/filter.rs

```diff
@@ -1195,46 +1196,105 @@ fn filter_internal(
     }
 }
 
+fn adaptive_filter(
+    f: impl Fn(&[u8]) -> u64,
+    bpp: usize,
+    len: usize,
+    previous: &[u8],
+    current: &[u8],
+    output: &mut [u8],
+) -> RowFilter {
+    use RowFilter::*;
+
+    let mut min_cost: u64 = u64::MAX;
+    let mut filter_choice = RowFilter::NoFilter;
+    for &filter in [Up, Sub, Avg, Paeth].iter() {
+        filter_internal(filter, bpp, len, previous, current, output);
+        let cost = f(output);
+        if cost <= min_cost {
+            min_cost = cost;
+            filter_choice = filter;
+
+            if cost == 0 {
+                return filter_choice;
+            }
+        }
+    }
+    if filter_choice != Paeth {
+        filter_internal(filter_choice, bpp, len, previous, current, output);
+    }
+    filter_choice
+}
+
 pub(crate) fn filter(
     method: Filter,
     bpp: BytesPerPixel,
     previous: &[u8],
     current: &[u8],
     output: &mut [u8],
 ) -> RowFilter {
-    use RowFilter::*;
     let bpp = bpp.into_usize();
     let len = current.len();
 
     match method {
-        Filter::Adaptive => {
-            let mut min_sum: u64 = u64::MAX;
-            let mut filter_choice = RowFilter::NoFilter;
-            for &filter in [Up, Sub, Avg, Paeth].iter() {
-                filter_internal(filter, bpp, len, previous, current, output);
-                let sum = sum_buffer(output);
-                if sum <= min_sum {
-                    min_sum = sum;
-                    filter_choice = filter;
-
-                    if sum == 0 {
-                        return filter_choice;
-                    }
-                }
-            }
-
-            if filter_choice != Paeth {
-                filter_internal(filter_choice, bpp, len, previous, current, output);
-            }
-            filter_choice
-        }
+        Filter::Adaptive => adaptive_filter(sum_buffer, bpp, len, previous, current, output),
+        Filter::MinEntropy => adaptive_filter(entropy, bpp, len, previous, current, output),
         _ => {
             let filter = RowFilter::from_method(method).unwrap();
             filter_internal(filter, bpp, len, previous, current, output)
         }
     }
 }
 
+/// Estimate the value of i * log2(i) without using floating point operations,
+/// implementation originally from oxipng.
+fn ilog2i(i: u32) -> u32 {
+    let log = 32 - i.leading_zeros() - 1;
+    i * log + ((i - (1 << log)) << 1)
+}
+
+fn entropy(buf: &[u8]) -> u64 {
+    let mut counts = [[0_u32; 256]; 4];
```

Comment:
Comment: Curious what the speed difference is here for this array being 1, 2, or 4 arrays large.

I believe you that it's faster, but slightly surprised that using 4 arrays and consolidating after performs better (I see that the consolidation loop is auto-vectorized when checked in Compiler Explorer).